ruby/lib/google/protobuf/descriptor_dsl.rb

```diff
@@ -0,0 +1,440 @@
+              # Update the default value in the proto.
+              field.default_value = name
+            end
+          else
+            # See if enum was defined in this file.
+            @file_proto.enum_type.each { |enum_proto|
+              if enum_proto.name == type_name
+                enum_proto.value.each { |enum_value_proto|
+                  if enum_value_proto.number == as_int
+                    # Update the default value in the proto.
+                    field.default_value = enum_value_proto.name
+                    return
+                  end
+                }
+                # We found the right enum, but no value matched.
+                return
+              end
+            }
+          end
+        end
+
+        # Historically we allowed enum defaults to be specified as a number.
+        # In retrospect this was a mistake as descriptors require defaults to
+        # be specified as a label. This can make a difference if multiple
+        # labels have the same number.
+        #
+        # Here we do a pass over all enum defaults and rewrite numeric defaults
+        # by looking up their labels.  This is complicated by the fact that the
+        # enum definition can live in either the symtab or the file_proto.
+        #
+        # We take advantage of the fact that this is called *before* enums or
+        # messages are nested in other messages, so we only have to iterate
+        # one level deep.
+        def rewrite_enum_defaults
+          @file_proto.message_type.each { |msg|
+            msg.field.each { |field|
+              rewrite_enum_default(field)
+            }
+          }
+        end
+
+        # We have to do some relatively complicated logic here for backward
+        # compatibility.
+        #
+        # In descriptor.proto, messages are nested inside other messages if that is
+        # what the original .proto file looks like.  For example, suppose we have this
+        # foo.proto:
+        #
+        # package foo;
+        # message Bar {
+        #   message Baz {}
+        # }
+        #
+        # The descriptor for this must look like this:
+        #
+        # file {
+        #   name: "test.proto"
+        #   package: "foo"
+        #   message_type {
+        #     name: "Bar"
+        #     nested_type {
+        #       name: "Baz"
+        #     }
+        #   }
+        # }
+        #
+        # However, the Ruby generated code has always generated messages in a flat,
+        # non-nested way:
+        #
+        # Google::Protobuf::DescriptorPool.generated_pool.build do
+        #   add_message "foo.Bar" do
+        #   end
+        #   add_message "foo.Bar.Baz" do
+        #   end
+        # end
+        #
+        # Here we need to do a translation where we turn this generated code into the
+        # above descriptor.  We need to infer that "foo" is the package name, and not
+        # a message itself. */
+
+        def get_parent_name(msg_or_enum)
```

Comment:
a getter that mutates it argument and returns  a value is fairly surprising.  Can we instead have this be

```
def split_parent_name(msg_or_enum)
  full_name = msg_or_enum.name
  idx = full.rindex(?.)
  if idx
    return (name[0...idx], name[idz+1..-1])
  else
    return nil, full_name
  end
end
```

Reply:
Done.